metaflow/plugins/kfp/kfp.py

```diff
@@ -998,23 +740,67 @@ def build_kfp_dag(
+                    + f" --flow_name {kfp_component.flow_name}"
+                    + f" --tags {json.dumps(json.dumps(kfp_component.tags))}"
+                    + f" --environment_type {kfp_component.environment_type}"
+                    + f" --logger_type {kfp_component.logger_type}"
+                    + f" --monitor_type {kfp_component.monitor_type}"
+                    + f" --user_code_retries {kfp_component.user_code_retries}"
+                    + " --workflow_name {{workflow.name}}"
+                    + f" --script_name {os.path.basename(sys.argv[0])}"
+                    + f" --passed_in_split_indexes={passed_in_split_indexes}"
+                    + f" --preceding_component_inputs={json.dumps(json.dumps(preceding_component_inputs))}"
+                    + f" --preceding_component_outputs={json.dumps(json.dumps(kfp_component.preceding_component_outputs))}",
+                ]
+
+                if node.name == "start":
+                    command[-1] += f" --flow_parameters_json='{flow_parameters_json}'"
+                if kfp_component.namespace:
+                    command[-1] += f" --namespace {kfp_component.namespace}"
+                if kfp_component.need_split_index:
+                    command[-1] += " --need_split_index"
+                for key in preceding_component_outputs_dict:
+                    command[-1] += f" {key}={preceding_component_outputs_dict[key]}"
+                command[-1] += ";c=$?; exit $c"
+
+                if (
+                    kfp_component.kfp_decorator
+                    and kfp_component.kfp_decorator.attributes["image"]
+                ):
+                    step_image = kfp_component.kfp_decorator.attributes["image"]
+                else:
+                    step_image = self.base_image
+
+                artifact_argument_paths = (
+                    None if node.name == "start" else {"flow_parameters_json": "None"}
                 )
-                container_op: ContainerOp = self.step_op(
-                    node.name,
-                    kfp_component,
-                    preceding_component_inputs=preceding_component_inputs,
-                    preceding_component_outputs=kfp_component.preceding_component_outputs,
-                )(**{**step_op_args, **preceding_component_outputs_dict})
+
+                file_outputs = {"foreach_splits": "/tmp/outputs/foreach_splits/data"}
+                for preceding_component_input in preceding_component_inputs:
+                    file_outputs[
+                        preceding_component_input
+                    ] = f"/tmp/outputs/{preceding_component_input}/data"
+
+                container_op = dsl.ContainerOp(
+                    name=node.name,
+                    image=step_image,
+                    command=command,
+                    artifact_argument_paths=artifact_argument_paths,
+                    file_outputs=file_outputs,
+                ).set_display_name(node.name)
```

Comment:
line 744 to here can be easily pulled out into a function

Reply:
I've pulled this out into a new function, `_generate_metaflow_execution_cmd()`.